Schema resource added in REST API Spec for api-version 2021-04-01-preview

shinjiDev opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [x] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  December 2021
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  December 2021
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [X] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [X] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [X] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [X] Check this box if any of the following apply to the PR so that label “WaitForARMFeedback” will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [X] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

solankisamir committed:
Adds base for updating Microsoft.ApiManagement from version preview/2021-01-01-preview to version 2021-04-01-preview

solankisamir committed:
Updates readme

solankisamir committed:
Updates API version in new specs and examples

azaslonov committed:
Added support for GraphQL API type (#14851)

* Added support for GraphQL API type

* Added missing enum value for GraphQL APIs.

Commit:
add schema definition fix from 2021-01-01-preview (#14883)

Co-authored-by: Alan Feng <alanfeng@microsoft.com>

solankisamir committed:
Vanguard: Get Outbound Network Dependency (#14900)

* outbound dependency

* fix resource

JatinSanghvi committed:
Update backup/restore API for managed identities (#14973)

* Update backup/restore API for system-assigned and user-assigned identities

* Fix formatting issue

* Rename user-assigned-msi-client-id with client-id

solankisamir committed:
SystemData implemenation on API Management Control Plane (#14899)

* system-data

* fix spec

* remove unreferenced file

atharos1 committed:
APIM - Network Watcher Connectivity Check integration specs (#15056)

* Connectivity Check API Specs

* Fix

* Fix errors

* style fixs

* Fix 202 code

* HTTPConnect request parameters

* Fix errors

Co-authored-by: Nicolás Barrera <t-nbarrera@microsoft.com>

VitaliyKurokhtin committed:
Parameter examples added (#14836)

solankisamir committed:
platformversion (#15114)

solankisamir committed:
fix update service (#15478)

atharos1 committed:
HttpConnect example (#15493)

* HttpConnect example

* Prettifier fix

Co-authored-by: Nicolás Barrera <t-nbarrera@microsoft.com>

Commit:
Rename from SchemaContract to ApiSchemaContract

Commit:
apimschema.json added to last apim preview version

Commit:
Completing specification for Schema resource.
Fixing changes in SchemaContract.

Commit:
Value field added to Schema resource specification.

solankisamir commented:
@shinjiDev can you fix the LintDiff and Spell Check errors in your PR

solankisamir commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/apimapis.json:
```diff
@@ -2143,7 +2143,7 @@
       }
     },
-    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/apis/{apiId}/schemas": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/apis/{apiId}/apischemas": {
```
I don't see we have changed the `ApiSchemaController`

maksimkim replied:
@solankisamir , feel free not to block merging 2021-04-01-preview into main on this PR. We'll request it separately.

shinjiDev replied:
@solankisamir It's going back to "schemas"

solankisamir commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/apimapis.json:
```diff
@@ -2204,7 +2204,7 @@
       }
     },
-    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/apis/{apiId}/schemas/{schemaId}": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/apis/{apiId}/apischemas/{apiSchemaId}": {
```
This requires change in the Backend also

shinjiDev replied:
It was undo

solankisamir reviewed (changes requested):
Let's postpone this to the next api-version. We are unnecessarily holding it hostage.

There are some decisions that need to be made
1. PUT should be long running
2. It should accept a Url Link instead of Inline Body, as ARM has 4MB limit.

Commit:
path fixed.
Renamed from ApiSchema to Schema.
Adding suppression rule to readme.md
Custom word added for spellcheck

maksimkim commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/apimanagement.json:
```diff
@@ -786,6 +786,16 @@
+      "in": "path",
+      "required": true,
+      "type": "string",
+      "description": "Schema entity identifier. Must be unique in the current API Management service instance. Must not have value 'managed'",
```
what's the purpose of not allowing "managed"?

maksimkim commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/apimanagement.json:
```diff
@@ -786,6 +786,16 @@
+      "description": "Schema entity identifier. Must be unique in the current API Management service instance. Must not have value 'managed'",
+      "minLength": 1,
+      "maxLength": 80,
+      "x-ms-parameter-location": "method"
```
why location is method?

maksimkim commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/examples/ApiManagementUpdateApiOperation.json:
```diff
@@ -62,7 +62,6 @@
             "representations": [
               {
                 "contentType": "text/xml",
-                "sample": "<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n<Envelope xmlns=\"http://schemas.xmlsoap.org/soap/envelope/\">\r\n  <Body>\r\n    <CancelOrder xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://tempuri.org/\">\r\n      <orderId>1</orderId>\r\n    </CancelOrder>\r\n  </Body>\r\n</Envelope>",
```
Why remove sample?

shinjiDev replied:
Because RepresentationContract in definitions.json doesn't have "sample" field in this preview.

maksimkim commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/apimanagement.json:
```diff
@@ -628,12 +628,12 @@
       "x-ms-parameter-location": "method"
     },
-    "SchemaIdParameter": {
+    "ApiSchemaIdParameter": {
```
Does it make sense to reuse SchemaIdParameter for both service level and api level schemas?

maksimkim commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/apimschema.json:
```diff
@@ -0,0 +1,329 @@
+            "in": "query",
+            "required": false,
+            "type": "string",
+            "description": "|     Field     |     Usage     |     Supported operators     |     Supported functions     |</br>|-------------|-------------|-------------|-------------|</br>| name | filter | ge, le, eq, ne, gt, lt | substringof, contains, startswith, endswith |</br>| region | filter | ge, le, eq, ne, gt, lt | substringof, contains, startswith, endswith |</br>| description | filter  ...
```
Let's not allow to filter by anything other than "name"

maksimkim commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/apimschema.json:
```diff
@@ -0,0 +1,329 @@
+            "$ref": "./examples/ApiManagementGetSchema2.json"
+          }
+        },
+        "produces": [
```
It's not going to produce anything but application/json. Schema type will be conveyed in schemaType property.

maksimkim commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/apimschema.json:
```diff
@@ -0,0 +1,329 @@
+            "$ref": "./apimanagement.json#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
```
Also add "202"

maksimkim commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/examples/ApiManagementCreateSchema1.json:
```diff
@@ -0,0 +1,39 @@
+    "parameters": {
+      "properties": {
+        "schemaType": "xml",
+        "value": "<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n<Envelope xmlns=\"http://schemas.xmlsoap.org/soap/envelope/\">\r\n  <Body>\r\n    <CancelOrder xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://tempuri.org/\">\r\n      <orderId>1</orderId>\r\n    </CancelOrder>\r\n  </Body>\r\n</Envelope>"
```
This seems to be an xml document. Could you add xml schema some simple xml schema instead?

maksimkim commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/examples/ApiManagementCreateSchema2.json:
```diff
@@ -0,0 +1,123 @@
+    "parameters": {
+      "properties": {
+        "schemaType": "json",
+        "document": {
```
This seems to be a swagger document. Could you add some simple json schema instead?

maksimkim commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/examples/ApiManagementGetApiOperation.json:
```diff
@@ -25,7 +25,6 @@
             "representations": [
               {
                 "contentType": "text/xml",
-                "sample": "<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n<Envelope xmlns=\"http://schemas.xmlsoap.org/soap/envelope/\">\r\n  <Body>\r\n    <CancelOrder xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://tempuri.org/\">\r\n      <orderId>1</orderId>\r\n    </CancelOrder>\r\n  </Body>\r\n</Envelope>",
```
Why removing sample?

shinjiDev replied:
Because RepresentationContract in definitions.json doesn't have "sample" field in this preview.

Commit:
Fixing PR observations.

Commit:
Added properties of difference API spec format for "SchemaDocumentProperties" (#15703)

* add schema definition fix from 2021-01-01-preview

* added properties for SchemaDocumentProperties

* updated the fix

* updated fix

RupengLiu committed:
[APIM]Add private endpoint connection APIs (#15115)

* Add private endpoint connection apis

* add readme

* quick fixes

* fix to PE contract

* Small fixes

* small fixes

* small fixes

* Small fixes

* small fixes

* small fix

* small fixes

* small fixes

* small fixes

* Update readme.md

* fix error

* small fix

* fix format

* address comments

* small fixes

* Change to lower case

* small fixes

Co-authored-by: msyyc <70930885+msyyc@users.noreply.github.com>

chiragg4u commented:
Please go through the contribution checklist added as part of the review and mark them completed. It will help you fix most of the issues and improve the quality of swagger and help close the PR faster. This is a requirement before review can be continued.

Commit:
Merge changes

Commit:
Merge main

Commit:
Modifications to Schema definition to add "description" field.
Updating examples of Schema.

Commit:
Merge from main

Commit:
Fixing apimanagement.json

Commit:
Prettier applied in files

Commit:
Fixing PR comments

Commit:
fixing x-ms-long-running-operation.

Commit:
x-ms-long-running-operation : true in apischema.

Commit:
Removing "Resource" type from apimanagement.json

shinjiDev commented:
> Let's postpone this to the next api-version. We are unnecessarily holding it hostage.
> 
> There are some decisions that need to be made
> 
> 1. PUT should be long running
> 2. It should accept a Url Link instead of Inline Body, as ARM has 4MB limit.

We got breaking changes approval.

shinjiDev reviewed (commented):
Thank you. It looks like everything is done.

zhenglaizhang commented:
@fengzhou-msft could you help check if we can ignore the .net sdk failure? 
@ArcturusZhang could you help check and appove go SDK breaking change? 
@msyyc could you help check python SDK breaking change?

maksimkim commented:
@zhenglaizhang is anything pending for this to be merged?

zhenglaizhang commented:
> @zhenglaizhang is anything pending for this to be merged?

this PR is adding new API which requires ARM signed off per https://github.com/Azure/azure-rest-api-specs/pull/15667#issuecomment-899869862 others LGTM.

shinjiDev commented:
> > @zhenglaizhang is anything pending for this to be merged?
> 
> this PR is adding new API which requires ARM signed off per [#15667 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/15667#issuecomment-899869862) others LGTM.

checkbox added.

mentat9 commented on specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-04-01-preview/apimschema.json:
```diff
@@ -0,0 +1,328 @@
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "./apimanagement.json#/definitions/ErrorResponse"
```
>ErrorResponse [](http://example.com/codeflow?start=57&length=13)

You can use definitions from types.json below common-types for these standard ARM properties (e.g. ErrorResponse, Resource): "$ref": "../../../../../common-types/resource-management/v3/types.json#/definitions/ErrorResponse".

shinjiDev replied:
Done

mentat9 commented:
Note: this PR is updating an existing API version rather than new service onboarding. Please update the submission form accordingly. I've reviewed the changes and left comments. Nothing is blocking for ARM signoff, but please consider the comments.

zhenglaizhang commented:
/azp run

shinjiDev committed:
Merge branch 'Azure:apim-2021-04-01-preview' into apim-2021-04-01-preview

Commit:
Apimschema.json updated to point to types.json for ErrorResponse.

shinjiDev commented:
> Note: this PR is updating an existing API version rather than new service onboarding. Please update the submission form accordingly. I've reviewed the changes and left comments. Nothing is blocking for ARM signoff, but please consider the comments.

PR description updated.

Commit:
undoing change of responseBody

Commit:
Merge branch 'apim-2021-04-01-preview' of https://github.com/shinjiDev/azure-rest-api-specs into apim-2021-04-01-preview

Commit:
Updating ErrorResponse.

Commit:
Fixing definitions merge.
